src/cargo/core/compiler/job_queue.rs

```diff
@@ -583,6 +583,17 @@ impl<'cfg> DrainState<'cfg> {
             }
         }
 
+        // If multiple pieces of work are waiting in the pending queue, we can sort it according to
+        // their priorities: higher priorities should be scheduled sooner.
+        if self.pending_queue.len() > 1 {
+            self.pending_queue.sort_by(|(unit_a, _), (unit_b, _)| {
```

Comment:
If we going to do a sort here anyway, then `self.queue.dequeue()` should not guarantee what order they items are returned in. In fact it should probably just return an iterator.

Also if we sort from lowest priority to highest priority, then when remove an item we can use the `.pop()` where we are now using `self.pending_queue.remove(0);`

Reply:
Yeah, this is why I also mentioned whether you wanted a different structure where the priority would be part of the data stored in the pending queue.

Ultimately it won't make a huge difference, especially with the allocation involved in caching, but those `Unit`s are indeed pretty big and their hashing may not be super cheap, so it should still improve a bit indeed.

I've pushed a change into a `sort_by_cached_key` call w/ `std::cmp::Reverse` since we need the order to be descending.

Reply:
@Eh2406 the sort will only apply if there are units in the pending queue here: so IIUC there can be cases (maybe when there are always jobserver tokens available) where the units could be processed one by one, in the order in which they are dequeued from the dependency queue. For these cases, we'd still need to keep the max-priority ordering in `self.queue.dequeue()`, right ?

Reply:
Yep. Using something like priority queue might be better. It is somehow a bit weird here
that we push the highest priority item from `self.queue.dequeue()`, and then immediately sort `pending_queue` in ascending order.

Anyway, we can always do it in a follow-up PR. I feel like we can merge it as-is. It would be wonderful if you can do another benchmark before r+. Thank you!

Reply:
Sure, what kind of benchmark would you like to see ?

Reply:
No particular one. The same benchmark you did is good enough. Just want to leave a trace for how much performance gains Cargo benefits from this commit.

Reply:
> It is somehow a bit weird here that we push the highest priority item from `self.queue.dequeue()`, and then immediately sort `pending_queue` in ascending order.

It is, and if we want to keep a Vec for now instead of a max-heap, we can insert the unit directly at the correct index instead.

> No particular one. The same benchmark you did is good enough

In [the report](https://github.com/lqd/rustc-benchmarking-data/tree/main/experiments/cargo-schedules/pending-queue-sorted) there are 800 benchmarks (× 7 runs × 2 versions of cargo × 3 parallelism settings × 2 build modes), so re-running _everything_ is not going to be possible soon unfortunately because it takes multiple days. I'll re-run some important ones, on a few different machines, e.g. tokio's `mini-redis` (I've been benchmarking that when pushing to this branch today, and it's still showing -9% at j4/6 for example) and some of the most positively and negatively affected crates.

I'll do those 2 things tomorrow.